Add YOLO9000 hierarchical softmax detections

relh opened the pull request:
## Summary

This PR adds YOLO9000-style hierarchical softmax support to Darkflow detection. It builds the WordTree and COCO mapping data needed by the YOLO9000 network, adds a `hierarchythreshold` option for deciding how far to traverse the label tree, and uses those outputs when producing detections.

It also adds the config data file paths needed by the existing YOLO9000 config and updates NMS9000 to suppress high-overlap boxes, which helped reduce duplicate body/head detections in sample outputs.

Fixes #49.

## Verification

- Ran sample data with `threshold=0.1` and `hierarchythreshold=0.6`; results were reasonable but not as strong as hoped.

relh committed:
YOLO 9000. Note new hierarchythreshold option variable

relh committed:
YOLO 9000. Note new hierarchythreshold option variable

relh committed:
YOLO9000. added new cython nms function and new
box_constructor, slight tweak to image tag formatting

abagshaw commented:
@relh Awesome! Can't wait to give this a shot when it's pulled. When I have some time I'll build a test for inference with YOLO9000 to increase coverage (right now it's failing as no tests cover your new code).

halt9 commented:
@relh I cloned your repo and gave it a shot, but here's what I get when I try to run the demo:
`Press [ESC] to quit demo
Traceback (most recent call last):
  File "/usr/local/bin/flow", line 6, in <module>
    cliHandler(sys.argv)
  File "/usr/local/lib/python2.7/dist-packages/darkflow/cli.py", line 25, in cliHandler
    tfnet.camera()
  File "/usr/local/lib/python2.7/dist-packages/darkflow/net/help.py", line 122, in camera
    single_out, img, False)
  File "/usr/local/lib/python2.7/dist-packages/darkflow/net/yolov2/predict.py", line 40, in postprocess
    boxes = self.findboxes(net_out)
  File "/usr/local/lib/python2.7/dist-packages/darkflow/net/yolov2/predict.py", line 28, in findboxes
    boxes=box_constructor9000(meta,net_out)
  File "darkflow/cython_utils/cy_yolo9000_findboxes.pyx", line 113, in darkflow.cython_utils.cy_yolo9000_findboxes.box_constructor (darkflow/cython_utils/cy_yolo9000_findboxes.c:3610)
    float hthreshold = meta['hierarchythreshold']
KeyError: 'hierarchythreshold'
`

This is after the initial setup phase.

abagshaw commented:
@halt9 Just a thought - I think darkflow is designed for Python 3 - and unless I'm mistaken, it appears you've installed it with Python 2. Could that be part of the problem?

relh commented:
I think it's a real issue that I don't initialize the hierarchythreshold in defaults. I'll actually test it with the demo rather than my external server I was running it with and update the PR!

abagshaw commented:
@relh I'm getting the same error on my system - is there a reason why you're accessing the `hierarchythreshold` value from `meta` and not `FLAGS`? As you defined it as a flag in `defaults.py` I would assume you'd be accessing it as a flag - no?

abagshaw commented:
@relh Thanks for your fix - that does seem to fix the issue, but now it removes the user's ability to change the value of `hierarchythreshold` for themselves in a flag.

I think the best/easiest way to fix this would just be to set the default value for `hierarchythreshold` in `defaults.py` to something like `0.6` like you suggested instead of `-0.1`.  Also might as well get rid of the condition: `if FLAGS.hierarchythreshold > 0.0:` in `darkflow/net/yolo/__init__.py` as that should always evaluate to true - so `self.meta['hierarchythreshold'] = FLAGS.hierarchythreshold` can always run. What do you think?

relh commented:
@abagshaw I realized that I think I was misusing meta, as I believe you guys have been only having meta store keys that are present in the cfg file. Since the hierarchy threshold isn't in the cfg (and hence meta), it makes more sense for it to be solely in flags. 

I agree adding it to defaults.py with a sensible default value makes sense: the key is just to make sure it's included as part of the parameters that get passed to the `box_constructor(meta,np.ndarray[float,ndim=3] net_out_in)` (so basically part of meta in some fashion)

I reverted it now to be similar to the darknet repo, where the value is hard coded in at 0.5.

I figured it would be worth talking about first because there is also a third threshold, the IOU threshold that is hard coded in darknet but could be another FLAGS parameter if we were interested in letting people control that. Adding the NMS IOU threshold however makes us have to add another parameter to the NMS function call..

So really the question becomes, how open is everyone to adding more to meta. I already added the 'hyponym_tree' and the 'coco_map' in the preprocessing stage to it.

abagshaw commented:
@relh Could we not just pass `FLAGS` to `box_constructor` as another argument? I'll let others chime in on how they think this should be best implemented.

On another note I've been playing around this for the past little while and just as you mentioned - detection accuracy isn't amazing. It appears to me that it's not even doing as well as YOLOv2 on the sample dataset. I'm going to try to compare these detections to the ones I get with the original darknet code to see if there are any noticeable differences. Maybe this is just the tradeoff you get with having tons of detection classes...

relh commented:
I'll add in these flags in a way that makes sense.

Beyond the flags, I'm planning to test out the original darknet repo for both the hierarchical 200 and 9000 class to see if the results from here are consistent. I know the 200 has existing detection examples online and it performs reasonably, but the real test is with the same images on the actual network.

halt9 commented:
Just wanted to point out the recent commits have indeed solved the issue that I was talking about above.

relh committed:
restructured new variables for meta to make sense, recombined cy_yolo9000 and cy_yolo2, added in hierarchical threshold

relh committed:
restructured new variables for meta to make sense, recombined cy_yolo9000 and cy_yolo2, added in hierarchical threshold

relh commented:
Ok, so I've restructured my PR.

Now, `net/yolo/__init__.py` adds the new keys to meta, similar to how the meta['labels'] key gets added. 

I also add the hierarchical threshold as per @abagshaw 's suggestion, taking it from the flags.

Beyond this I've removed the separate `cy_yolo9000_findboxes.pyx` file as it wasn't necessary and put the added functionality back in `cy_yolo2_findboxes.pyx`.

Currently it checks the meta to see if certain keys exist to determine what kind of object detection to perform, this means it defaults to the softmax across the 200 items as described in the paper with the yolo9000 weights as both the `tree` and `map` meta keys are present.

It might make more sense to provide a configuration option to choose between yolo2/yolo200/yolo9000. I'd hope for some input from you all on this one.

It is also worth noting that in this thread https://groups.google.com/forum/#!topic/darknet/Cw99SpxKK2A they comment on the not excellent/improved accuracy of the 9000 network.

I ran darknet for comparisons against my results and didn't find many differences (see https://github.com/philipperemy/yolo-9000/issues/3).

I'm still noting some weirdness with the 200 class version in one of the demo images with a horse categorizing as an apple rather than a horse.

@halt9 Are you using the default threshold values? if you are running flow to test with a video, you won't see many bounding boxes unless you use a lower threshold. Here is a command I ran with a demo video I had:

`flow --model cfg/yolo9000.cfg --load bin/yolo9000.weights --threshold 0.1 --hierarchyThreshold 0.1 --demo videofile.avi` 

I've got to leave this alone for a few days for work, but think it's getting closer to joining the main repository.

cjosephson commented:
What is the status on this? Would love to see it added
